embassy-stm32/src/usart/mod.rs

```diff
@@ -1466,6 +1483,120 @@ fn reconfigure(info: &Info, kernel_clock: Hertz, config: &Config) -> Result<(),
     Ok(())
 }
 
+fn calculate_brr(baud: u32, pclk: u32, presc: u32, mul: u32) -> u32 {
+    // The calculation to be done to get the BRR is `mul * pclk / presc / baud`
+    // To do this in 32-bit only we can't multiply `mul` and `pclk`
+    let clock = pclk / presc;
+
+    // The mul is applied as the last operation to prevent overflow
+    let brr = clock / baud * mul;
+
+    // The BRR calculation will be a bit off because of integer rounding.
+    // Because we multiplied our inaccuracy with mul, our rounding now needs to be in proportion to mul.
+    let rounding = ((clock % baud) * mul + (baud / 2)) / baud;
+
+    brr + rounding
+}
+
+fn set_baudrate(info: &Info, kernel_clock: Hertz, baudrate: u32) -> Result<(), ConfigError> {
+    info.interrupt.disable();
+
+    set_usart_baudrate(info, kernel_clock, baudrate)?;
+
+    info.interrupt.unpend();
+    unsafe { info.interrupt.enable() };
+
+    Ok(())
+}
+
+fn set_usart_baudrate(info: &Info, kernel_clock: Hertz, baudrate: u32) -> Result<(), ConfigError> {
```

Comment:
Sure I can, but my only concern is that calling `configure` would require the developer to pass the Config struct again to avoid overwriting any previous configuration made on Usart instance construction when perhaps all it wants is just to change the `baudrate` all this since Config is not copied in Usart instance to reuse its values.

I noticed I can reduce code duplication by moving `static DIVS`  to the outer scope, and wrap Usart `kind` matching and `brr` search in two separate functions to be used by both `configure` and `set_usart_baudrate` if you think this would be better.

Reply:
`configure` does more things, so I meant make `configure()` call `set_usart_baudrate()` then do the other stuff.

Reply:
Ah, sorry I misunderstood you. But yeah that makes total sense. I just pushed changes addressing this. Calling `find_and_set_brr` in both `configure` and `set_baud_rate`  reducing code duplication. Please take a look and let me know what you think.